linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala

```diff
@@ -1076,60 +1000,68 @@ private[emitter] object CoreJSLib {
           }
         }
 
-        if (useECMAScript2015) {
-          val f = weakMapBasedFunction
-          defineFunction("systemIdentityHashCode", f.args, f.body)
-        } else {
-          buf += extractWithGlobals(globalVarDef("systemIdentityHashCode", CoreVar,
-              If(idHashCodeMap !== Null(), weakMapBasedFunction, fieldBasedFunction)))
-        }
+        Block(
+            let(lastIDHash, 0),
+            const(idHashCodeMap,
+                if (useECMAScript2015) New(WeakMapRef, Nil)
+                else If(typeof(WeakMapRef) !== str("undefined"), New(WeakMapRef, Nil), Null())),
+            if (useECMAScript2015) {
+              val f = weakMapBasedFunction
+              defineFunction("systemIdentityHashCode", f.args, f.body)
+            } else {
+              extractWithGlobals(globalVarDef("systemIdentityHashCode", CoreVar,
+                  If(idHashCodeMap !== Null(), weakMapBasedFunction, fieldBasedFunction)))
+            }
+        )
       }
-    }
+    )
 
-    private def defineIsPrimitiveFunctions(): Unit = {
-      val v = varRef("v")
+    private def defineIsPrimitiveFunctions(): Tree = Block(
+      locally {
+        val v = varRef("v")
 
-      def defineIsIntLike(name: String, specificTest: Tree): Unit = {
-        defineFunction(name, paramList(v), {
-          Return((typeof(v) === str("number")) && specificTest &&
-              ((int(1) / v) !== (int(1) / double(-0.0))))
-        })
-      }
+        def defineIsIntLike(name: String, specificTest: Tree): Tree = {
+          defineFunction(name, paramList(v), {
```

Comment:
```suggestion
          defineFunction1(name) { v =>
```
and therefore remove the `val v`?

Reply:
I'll need to change `specificTest` to a `VarRef => Tree`, but I think that is indeed better.